api/dataprocessor.go

```diff
@@ -124,119 +138,136 @@ func (s *Server) getTargets(ctx context.Context, reqs []models.Req) ([]models.Se
+			buf, err := node.Post(rCtx, "getTargetsRemote", "/getdata", models.GetData{Requests: reqs})
 			if err != nil {
-				errorsChan <- err
+				cancel()
+				responses <- getTargetsResp{nil, err}
 				return
 			}
 			var resp models.GetDataResp
 			_, err = resp.UnmarshalMsg(buf)
 			if err != nil {
+				cancel()
 				log.Error(3, "DP getTargetsRemote: error unmarshaling body from %s/getdata: %q", node.Name, err)
-				errorsChan <- err
+				responses <- getTargetsResp{nil, err}
 				return
 			}
 			log.Debug("DP getTargetsRemote: %s returned %d series", node.Name, len(resp.Series))
-			seriesChan <- resp.Series
-		}(ctx, nodeReqs)
+			responses <- getTargetsResp{resp.Series, nil}
+		}(nodeReqs)
 	}
+
+	// wait for all getTargetsRemote goroutines to end, then close our responses channel
 	go func() {
 		wg.Wait()
-		close(seriesChan)
-		close(errorsChan)
+		close(responses)
 	}()
+
 	out := make([]models.Series, 0)
-	var err error
-	for series := range seriesChan {
-		out = append(out, series...)
+	for resp := range responses {
+		if resp.err != nil {
+			return nil, resp.err
+		}
+		out = append(out, resp.series...)
 	}
 	log.Debug("DP getTargetsRemote: total of %d series found on peers", len(out))
-	for e := range errorsChan {
-		err = e
-		break
-	}
-	return out, err
+	return out, nil
 }
 
 // error is the error of the first failing target request
 func (s *Server) getTargetsLocal(ctx context.Context, reqs []models.Req) ([]models.Series, error) {
 	log.Debug("DP getTargetsLocal: handling %d reqs locally", len(reqs))
-	seriesChan := make(chan models.Series, len(reqs))
-	errorsChan := make(chan error, len(reqs))
-	// TODO: abort pending requests on error, maybe use context, maybe timeouts too
-	wg := sync.WaitGroup{}
-	wg.Add(len(reqs))
+	responses := make(chan getTargetsResp, len(reqs))
+
+	var wg sync.WaitGroup
+	reqLimiter := newLimiter(getTargetsConcurrency)
+
+	rCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+LOOP:
 	for _, req := range reqs {
-		go func(ctx context.Context, wg *sync.WaitGroup, req models.Req) {
-			ctx, span := tracing.NewSpan(ctx, s.Tracer, "getTargetsLocal")
+		// check to see if the request has been canceled, if so abort now.
+		select {
+		case <-rCtx.Done():
+			//request canceled
+			break LOOP
+		default:
+		}
+		// if there are already getDataConcurrency goroutines running, then block
+		// until a slot becomes free.
+		reqLimiter.enter()
+		wg.Add(1)
+		go func(req models.Req) {
+			rCtx, span := tracing.NewSpan(rCtx, s.Tracer, "getTargetsLocal")
 			req.Trace(span)
-			defer span.Finish()
 			pre := time.Now()
-			points, interval, err := s.getTarget(ctx, req)
+			points, interval, err := s.getTarget(rCtx, req)
 			if err != nil {
 				tags.Error.Set(span, true)
-				errorsChan <- err
+				cancel() // cancel all other requests.
+				responses <- getTargetsResp{nil, err}
 			} else {
 				getTargetDuration.Value(time.Now().Sub(pre))
```

Comment:
if getTarget() returned early due to a cancel, than we will report an incorrect (and possibly unrealistically low) getTargetDuration here

Reply:
@woodsaj ^^ ?

Reply:
The metric will correctly reflect the time spent running getTarget(). 
Is getTargetDuration meant to represent something else?

Reply:
yes. the metric represents the getting of a target. if getTarget is cancelled due to whatever reason and is not actually getting targets, then we shouldn't report this duration as it does not apply.